Factor out `CompilationUnit` definitions into implementation file

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59465
* __->__ #59245
* #59244

Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

Commit:
Factor out `CompilationUnit` definitions into implementation file

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

malfet commented:
Please add a new file to `core_sources_full_mobile` in `build_variables.bzl`
https://github.com/pytorch/pytorch/blob/8805093ec5c79f689824b6169ddc81bd44067dff/tools/build_variables.bzl#L158

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/api/compilation_unit.cpp:
```diff
@@ -0,0 +1,373 @@
+#include <torch/csrc/jit/api/compilation_unit.h>
```
Thanks for making our implementation more organized!

gmagogsfm commented on torch/csrc/jit/api/compilation_unit.cpp:
```diff
@@ -0,0 +1,373 @@
+#include <torch/csrc/jit/api/compilation_unit.h>
+#include <torch/csrc/jit/frontend/ir_emitter.cpp>
```
Curious about the decision to include `cpp` file. Doing so can invite unwanted side effects.

https://www.cplusplus.com/forum/general/39618/

ansley replied:
It was very difficult to get the headers right, and I realized that everything Just Worked™ when I included the `cpp` file. I didn't know there were such serious consequences. Thanks for the link!

Does this also apply to tests? I have a `cpp` file in `test_cleanup_passes.cpp`. If I use the header file instead, then the implementation isn't linked by the time the test is compiled. If I put the definition in the header, then there's a problem with `RunCleanupPasses` being defined twice

SplitInfinity replied:
> I have a cpp file in `test_cleanup_passes.cpp`. If I use the header file instead, then the implementation isn't linked by the time the test is compiled.

This suggests that some change needs to be made in some `CMakeLists.txt` to add the new dependency. 🤔

ansley replied:
Hmm, I'd assume that `run_cleanup_passes.cpp` would need to be specified in `build_variables.bzl::core_sources_full_mobile`, but that doesn't seem to be doing it. 

cc @malfet

ansley replied:
Resolved by adding `TORCH_API`

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3360,370 +3360,372 @@ struct to_ir {
-  // calles the define with native resolver to generate the graph for functions
-  define(c10::nullopt, source, nativeResolver(), nullptr);
-}
+//CompilationUnit::CompilationUnit(const std::string& source)
```
Should we delete code instead of commenting them out?

ansley replied:
Oops, thanks for catching this!

SplitInfinity commented on tools/build_variables.bzl:
```diff
@@ -155,7 +155,7 @@ core_sources_full_mobile = [
     "torch/csrc/jit/frontend/exit_transforms.cpp",
     "torch/csrc/jit/frontend/inline_loop_condition.cpp",
-    "torch/csrc/jit/frontend/ir_emitter.cpp",
+    "torch/csrc/jit/api/compilation_unit.cpp",
```
While we're moving code around, what are people's opinions on keeping `CompilationUnit` in the `api` folder? It is part of the public API, but not to the same extent as `Module`, `Object`, and even `Method`.

SplitInfinity reviewed (commented):
Thanks for doing this. I think it will be ready to land once the instance of `#include` with a `.cpp` is fixed and the commented out code in `ir_emitter.cpp` is removed.

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

gmagogsfm commented on test/cpp/jit/test_constant_pooling.cpp:
```diff
@@ -93,7 +93,7 @@ graph():
 
 // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-TEST(constantPoolingTest, DictConstantPooling) {
+TEST(constantPoolingTest, DictconstantPooling) {
```
Accident? It seems DictConstantPooling is correct CamelCase.

ansley replied:
Good catch!

gmagogsfm reviewed (approved):
Looks good. Just one minor comment and some tests (due to merge conflicts) need to be fixed.

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

Commit:
Update on "Factor out `CompilationUnit` definitions into implementation file"


Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).
